hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/IOmMetadataReader.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.hadoop.ozone.om;
+
+import org.apache.hadoop.ozone.OzoneAcl;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.helpers.OmKeyArgs;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.OzoneFileStatus;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Protocol for OmMetadataReader's.
+ */
+public interface IOmMetadataReader {
+  /**
+   * Look up for the container of an existing key.
+   *
+   * @param args the args of the key.
+   * @return OmKeyInfo instance that client uses to talk to container.
+   */
+  OmKeyInfo lookupKey(OmKeyArgs args) throws IOException;
+
+  /**
+   * List the status for a file or a directory and its contents.
+   *
+   * @param args    Key args
+   * @param recursive  For a directory if true all the descendants of a
+   *                   particular directory are listed
+   * @param startKey   Key from which listing needs to start. If startKey exists
+   *                   its status is included in the final list.
+   * @param numEntries Number of entries to list from the start key
+   * @param allowPartialPrefixes if partial prefixes should be allowed,
+   *                             this is needed in context of ListKeys
+   * @return list of file status
+   */
+  List<OzoneFileStatus> listStatus(OmKeyArgs args, boolean recursive,
```

Comment:
it works for the FS interface, but doesnt seem to work for object store interface. This is what I see on my setup

bash-4.2$ ozone fs -ls /vol1/bucket1/.snapshot/snap1
Found 1 items
-rw-rw-rw-   1 hadoop hadoop       3811 2022-09-07 21:05 /vol1/bucket1/.snapshot/snap1/k1

bash-4.2$ ozone sh key list /vol1/bucket1/.snapshot/snap1
Invalid value for positional parameter at index 0 (<value>): cannot convert '/vol1/bucket1/.snapshot/snap1' to OzoneAddress (org.apache.hadoop.ozone.client.OzoneClientException: Invalid bucket name. Delimiters (/) not allowed in bucket name)

Reply:
When you mentioned this issue earlier today, I didn't realize you meant this command.  I have seen it before. It is not really a snapshot issue but a UI issue.  I was going to add a separate jira ticket for it.

This is actually a bucket level command and requires the parameter to be a valid bucket name, (as you can see from the error message: "Invalid bucket name. Delimiters (/) not allowed in bucket name").  It is easy enough to fix, we just have to decide what the UI should be.

Two possible options are:
1. Change the client side bucket name validation to allow snapshot names, (which seems a bit ugly to me.)
2. Allow as separate optional --snapshotName parameter for this UI.

Reply:
We can do both 1 & 2. Anyone who wants to us ".snapshot" syntax can do that. Optionally, bucket/Key level commands can provide --snapshot parameter too. This could be just a UI side fix. On the server side, we should get the snapshot name as a proto parameter.

Reply:
Actually, after responding to @smengcl 's comment above about about the listKeys() keyPrefix, I realized that this command does work, but has to be invoked with a keyPrefix.  So, instead of this:
```
ozone sh key list /vol1/bucket1/.snapshot/snap1
```
It works like this:
```
ozone sh key list /vol1/bucket1/ -p .snapshot/snap1
```

I admit that is not a great user interface, and should probably be fixed.  But at least you can see the functionality is there.